File: clang-codes/record/record.cpp
/*
Get information about typedef, struct/union/class/enum
format:
//if enum:
    <enum>
    enum_name
    sourcerange
    num fields(1, 2, ...N)
    field1 name
    field2 name
    ...
    fieldN name
//if typedef:
    <typedef>
    typedef name
    underlying type
    source range
//if record(struct/union/class):
    <struct/union/class>
    name
    sourcerange
    num fields(1, 2, ...N)
    field1_name type
    field2_name type
    ...
    fieldN_name type
    <Record>(c)/<CXXRecord>(c++)
    if CXXRecord:
        num methods(1, 2, ...N)
        <constructor>
        ...
        <constructor>
        <destructor>
        ...
        <destructor>
        <method>
        ...
        .
        .
        .
        ...
        <method>
        * ... -->
        name
        sourcerange
        storage_container static/const/volatile/virtual
        num parameters(1, 2, ...N)
        return type
        param1 type
        param2 type
        ...
        paramN type

*/

#include "llvm/Support/CommandLine.h"

#include "clang/Basic/SourceManager.h"
#include "clang/Basic/FileManager.h"

#include "clang/Frontend/TextDiagnosticPrinter.h"
#include "clang/Frontend/CompilerInstance.h"
#include "clang/Frontend/FrontendActions.h"
#include "clang/Frontend/FrontendOptions.h"
#include "clang/Frontend/Utils.h"

#include "clang/Lex/Lexer.h"
#include "clang/Lex/Token.h"

#include "clang/Tooling/CommonOptionsParser.h"
#include "clang/Tooling/Tooling.h"

#include "clang/AST/RecursiveASTVisitor.h"
#include "clang/AST/ASTConsumer.h"
#include "clang/AST/ASTContext.h"
#include "clang/AST/ASTTypeTraits.h"

#include <memory>
#include <vector>
#include <iostream>


using namespace clang;
using namespace clang::tooling;

static llvm::cl::OptionCategory MyOpts("Ignored");

std::string fileName;

// print storage container type
#define stringify( name ) # name
const char* SC[] = 
  {
    stringify(SC_None),
    stringify(SC_Extern),
    stringify(SC_Static),
    stringify(SC_PrivateExtern),
    stringify(SC_Auto),
    stringify(SC_Register)
  };

class customVisitor
  : public RecursiveASTVisitor<customVisitor> {
public:
  explicit customVisitor(ASTContext *Context)
    : Context(Context) {}
    
    bool VisitDecl(Decl *Declaration) {
        
        clang::SourceManager& mgr = Context->getSourceManager();
        if (mgr.isInMainFile(Declaration->getLocation())) {
            //enum
            if(!strcmp(Declaration->getDeclKindName(), "Enum")) {
                EnumDecl *enumDecl = llvm::dyn_cast<EnumDecl>(Declaration);
                if (enumDecl->isCompleteDefinition()) {
                    std::cout<<"<"<<enumDecl->getKindName().str()<<">"<<"\n";
                    std::cout<<enumDecl->getNameAsString()<<"\n";
                    // sourcerange
                    SourceRange srcRange = enumDecl->getSourceRange();
                    if (srcRange.isValid()) {
                        std::cout<<srcRange.printToString(mgr)<<"\n";
                    }
                    else {
                        std::cout<<"invalid SourceRange\n";
                    }
                    //print enum fields
                    std::string fields = "";
                    int totalFields = 0;
                    for(auto enum_iter= enumDecl->enumerator_begin(); enum_iter!=enumDecl->enumerator_end(); enum_iter++) {
                        totalFields += 1;
                        fields = fields + enum_iter->getNameAsString() + "\n";
                        //std::cout<<enum_iter->getNameAsString()<<"\n";
                    }
                    std::cout<<totalFields<<'\n'<<fields;
                }
            }
            //struct, union, class
            else if(!strcmp(Declaration->getDeclKindName(), "Record") || !strcmp(Declaration->getDeclKindName(), "CXXRecord")) {
                RecordDecl *record = llvm::dyn_cast<RecordDecl>(Declaration);
                if (record->isCompleteDefinition()) {
                    std::cout<<"<"<<record->getKindName().str()<<">"<<"\n";
                    std::string recordName = record->getNameAsString();
                    std::cout<<recordName<<"\n";
                    // sourcerange
                    SourceRange srcRange = record->getSourceRange();
                    if (srcRange.isValid()) {
                        std::cout<<srcRange.printToString(mgr)<<"\n";
                    }
                    else {
                        std::cout<<"invalid SourceRange\n";
                    }
                    // fields
                    std::string fields = "";
                    int totalFields = 0;
                    for(auto field_iter= record->field_begin(); field_iter!=record->field_end(); field_iter++) {
                        totalFields += 1;
                        fields = fields + field_iter->getNameAsString() + " " + field_iter->getType().getAsString() + "\n";
                        //std::cout<<field_iter->getNameAsString()<<" "<<field_iter->getType().getAsString()<<"\n";
                    }
                    std::cout<<totalFields<<'\n'<<fields;
                    std::cout<<Declaration->getDeclKindName()<<"\n";
                    //cpp record methods
                    if(!strcmp(Declaration->getDeclKindName(), "CXXRecord")) {
                        //cast to CXXRecordDecl
                        CXXRecordDecl *cxxRecord = llvm::dyn_cast<CXXRecordDecl>(Declaration);
                        //methods
                        std::string methods = "";
                        std::string constructor = "";
                        std::string destructor = "";
                        std::string destStartStr = "~";
                        int totalMethods = 0;
                        for(auto method_iter = cxxRecord->method_begin(); method_iter != cxxRecord->method_end(); method_iter++) {
                            //check method user defined
                            if(method_iter->isUserProvided()) {
                                totalMethods += 1;
                                //method name
                                std::string methodName = method_iter->getNameAsString();
                                std::string method = methodName + "\n";
                                //source range
                                SourceRange methodSrcRange = method_iter->getSourceRange();
                                if (methodSrcRange.isValid()) {
                                    method += methodSrcRange.printToString(mgr) + "\n";
                                }
                                else {
                                    method += "invalid SourceRange\n";
                                }

                                method += std::string(SC[method_iter->getStorageClass()]) +" ";
                                
                                if(method_iter->isStatic()) {
                                    method += "static ";
                                }
                                //if(method_iter->isInstance()) {
                                //    method += "instance ";
                                //}
                                if(method_iter->isConst()) {
                                    method += "const ";
                                }
                                if(method_iter->isVolatile()) {
                                    method += "volatile ";
                                }
                                if(method_iter->isVirtual()) {
                                    method += "virtual ";
                                }
                                //return type
                                method += method_iter->getReturnType().getAsString() +"\n";
                                //parameters
                                int numParams = method_iter->getNumParams();
                                method = method + std::to_string(numParams) + "\n";
                                
                                for(int i=0; i<numParams; i++) {
                                    auto param = method_iter->parameters()[i];
                                    method = method + param->getNameAsString() + " " + param->getOriginalType().getAsString() + "\n";
                                }
                                method += "\n";
                                //constructor or destructor or
                                if(methodName==recordName) {
                                    constructor += method;
                                }
                                else if(methodName == (destStartStr+recordName)) {
                                    destructor += method;
                                }
                                else {
                                    methods += method;
                                }
                            }
                        }
                        std::cout<<totalMethods<<"\n";
                        if(totalMethods) {
                            std::cout<<"<constructor>\n"<<constructor<<"<constructor>\n";
                            std::cout<<"<destructor>\n"<<destructor<<"<destructor>\n";
                            std::cout<<"<method>\n"<<methods<<"<method>\n";
                        }
                    }
                }
            }
            else if(!strcmp(Declaration->getDeclKindName(), "Typedef")) {
                TypedefDecl *Typedef = llvm::dyn_cast<TypedefDecl>(Declaration);
                std::cout<<"<typedef>\n";
                std::cout<<Typedef->getNameAsString()<<"\n"<<Typedef->getUnderlyingType().getAsString()<<"\n";
                SourceRange srcRange = Typedef->getSourceRange();
                if (srcRange.isValid()) {
                    std::cout<<srcRange.printToString(mgr)<<"\n";
                }
                else {
                    std::cout<<"invalid SourceRange\n";
                }
            }

        }

        return true;

    }

private:
    ASTContext *Context;
};

class TraverseAstConsumer : public clang::ASTConsumer {
public:
    explicit TraverseAstConsumer(ASTContext *Context)
        : Visitor(Context) {}

    virtual void HandleTranslationUnit(clang::ASTContext &Context) {
        Visitor.TraverseDecl(Context.getTranslationUnitDecl());
    }
private:
    customVisitor Visitor;
};

class TraverseAstAction : public clang::ASTFrontendAction {
public:
    virtual std::unique_ptr<clang::ASTConsumer> CreateASTConsumer(
    clang::CompilerInstance &Compiler, llvm::StringRef InFile) {

        return std::unique_ptr<clang::ASTConsumer>(
            new TraverseAstConsumer(&Compiler.getASTContext()));
    }
};

int main(int argc, const char ** argv)
{
    fileName = argv[1];
    CommonOptionsParser opt_prs(argc, argv, MyOpts);
    ClangTool tool(opt_prs.getCompilations(), opt_prs.getSourcePathList());
    int result = tool.run(newFrontendActionFactory<TraverseAstAction>().get());
    return result;
}

File: python/record.py
from identifier import getLocation, function

#get information from enum declaration
def enum(lines):
    #first line is the name of the enum
    enumName = lines[0]
    #second line is the sourceRange
    srcRange = lines[1]
    if srcRange == "invalid SourceRange":
        location = []
    else:
        location = getLocation(srcRange)
    #third line is number of fields followed by name of fields
    fields = []
    numFields = int(lines[2])
    for i in range(numFields):
        fields.append(lines[i+3])
    return (enumName, location, fields)

#get information from typedef declaration
def typedef(lines):
    #first line is the name of the typedef
    typedefName = lines[0]
    #second line is underlying type
    undType = lines[1].split()
    #third line is the sourceRange
    srcRange = lines[2]
    if srcRange == "invalid SourceRange":
        location = []
    else:
        location = getLocation(srcRange)
    return (typedefName, undType, location)

#get information about method
#format (same as function except methods are separated by newline)
### name
### sourcerange
### storage_container static/const/volatile/virtual
### num parameters(1, 2, ...N)
### return type
### param1 type
### param2 type
### ...
### paramN type
### <blank>
def method(lines):
    length = len(lines)
    idx = 0
    unit = []
    methods = {}
    while idx<length:
        if len(lines[idx]):
            unit.append(lines[idx])
        else:#new method
            method_ = function(unit)
            methodName = method_[0]
            if methods.get(methodName) is None:
                methods[methodName] = []
            methods[methodName].append( {
                    'location':method_[1], 'return':method_[2],
                    'num':method_[3], 'params':method_[4]
                } )
            unit = []
        idx += 1
    return methods

#return constructor/destructor/method lines
def methodUnits(lines, methodType):
    unit = []
    #units between <constructor>...<constructor> or destructor or method
    assert lines[0]==methodType
    idx = 1
    while lines[idx]!=methodType:
        unit.append(lines[idx])
        idx += 1
    return unit, idx

#get information from record declaration
def record(lines):
    length = len(lines)
    #first line is the name
    recordName = lines[0]
    #second line is the sourceRange
    srcRange = lines[1]
    if srcRange == "invalid SourceRange":
        location = []
    else:
        location = getLocation(srcRange)
    #third line is number of fields followed by name of fields
    fields = {}
    numFields = int(lines[2])
    for i in range(numFields):
        currField = lines[i+3].split()
        if fields.get(currField[0]) is None:
            fields[currField[0]] = []
        fields[currField[0]].append(currField[1:])
    #next line Record for c records; CXXRecord for c++
    idx = len(fields)+3
    methods = {}
    if lines[idx]=="CXXRecord":
        methodLines = lines[idx+1:]
        numMethods = int(methodLines[0])
        if numMethods:
            methodLines = methodLines[1:]
            # first constructor unit
            unit, idx = methodUnits(methodLines, '<constructor>')
            methods['constructor'] = method(unit)
            #next unit
            methodLines = methodLines[idx+1:]
            # second destructor unit
            unit, idx = methodUnits(methodLines, '<destructor>')
            methods['destructor']  = method(unit)
            #next unit
            methodLines = methodLines[idx+1:]
            # last unit other methods
            unit, idx = methodUnits(methodLines, '<method>')
            methods['method']  = method(unit)
    
    return (recordName, location, fields, methods)

'''
Get information about typedef, struct/union/class/enum
format:
//if enum:
    <enum>
    enum_name
    sourcerange
    num fields(1, 2, ...N)
    field1 name
    field2 name
    ...
    fieldN name
//if typedef:
    <typedef>
    typedef name
    underlying type
    source range
//if record(struct/union/class):
    <struct/union/class>
    name
    sourcerange
    num fields(1, 2, ...N)
    field1_name type
    field2_name type
    ...
    fieldN_name type
    <Record>(c)/<CXXRecord>(c++)
    if CXXRecord:
        num methods(1, 2, ...N)
        <constructor>
        ...
        <constructor>
        <destructor>
        ...
        <destructor>
        <method>
        ...
        .
        .
        .
        ...
        <method>
    * ... -->
    ~~~~~~~~~~
        name
        sourcerange
        storage_container static/const/volatile/virtual
        num parameters(1, 2, ...N)
        return type
        param1 type
        param2 type
        ...
        paramN type
        
    ~~~~~~~~~~
'''

def getRecordData(lines):
    unit = []
    enums = {}
    typedefs = {}
    records = {}
    length = len(lines)
    if length==0:
        return enums, typedefs, records
    #record kinds
    kinds = ["<enum>", "<typedef>", "<struct>", "<union>", "<class>"]
    #flag list for checking unit which kind; idx 0 enum, idx 1 typedef, idx 2 others
    flag = [False, False, False]
    #Record/typedef to flag list index map
    map_ = {
        kinds[0] : 0,
        kinds[1] : 1,
    }
    for i in range(2,5):
        map_[kinds[i]] = 2
    if lines[0] in kinds:
        flag[map_[lines[0]]] = True
    else:
        print("Error Record Format")
        return None, None, None
    idx=1
    while idx<length:
        # new declaration begins
        if lines[idx] in kinds:
            if flag[0]:#prev decl is enum
                enum_ = enum(unit)
                enumName = enum_[0]
                if enums.get(enumName) is None:
                    enums[enumName] = []
                enums[enumName].append({
                    'location' : enum_[1],
                    'fields' : enum_[2]
                })
            elif flag[1]:#prev decl is typedef
                typedef_ = typedef(unit)
                typedefName = typedef_[0]
                if typedefs.get(typedefName) is None:
                    typedefs[typedefName] = []
                typedefs[typedefName].append({
                    'type' : typedef_[1],
                    'location' : typedef_[2]
                })
            elif flag[2]:
                recordType = lines[idx-len(unit)-1][1:-1]#struct/union/class
                record_ = record(unit)
                recordName = record_[0]
                if records.get(recordName) is None:
                    records[recordName] = []
                records[recordName].append({
                    'type' : recordType,
                    'location' : record_[1],
                    'fields' : record_[2],
                    'methods' : record_[3]
                })
            flag = [False, False, False]
            flag[map_[lines[idx]]] = True
            unit = []
        # continue with current declaration
        else:
            unit.append(lines[idx])
        idx += 1
    #last declaration
    if flag[0]:#prev decl is enum
        enum_ = enum(unit)
        enumName = enum_[0]
        if enums.get(enumName) is None:
            enums[enumName] = []
        enums[enumName].append({
            'location' : enum_[1],
            'fields' : enum_[2]
        })
    elif flag[1]:#prev decl is typedef
        typedef_ = typedef(unit)
        typedefName = typedef_[0]
        if typedefs.get(typedefName) is None:
            typedefs[typedefName] = []
        typedefs[typedefName].append({
            'type' : typedef_[1],
            'location' : typedef_[2]
        })
    elif flag[2]:
        recordType = lines[idx-len(unit)-1][1:-1]#struct/union/class
        record_ = record(unit)
        recordName = record_[0]
        if records.get(recordName) is None:
            records[recordName] = []
        records[recordName].append({
            'type' : recordType,
            'location' : record_[1],
            'fields' : record_[2],
            'methods' : record_[3]
        })
    return enums, typedefs, records
